Snowflake.Data/Core/FileTransfer/SFFileTransferAgent.cs

```diff
@@ -185,46 +185,63 @@ class SFFileTransferAgent
+                // Initialize the encryption metadata
+                initEncryptionMaterial();
 
-                Directory.CreateDirectory(TransferMetadata.localLocation);
-            }
+                if (CommandTypes.UPLOAD == CommandType)
+                {
+                    initFileMetadataForUpload();
+                }
+                else if (CommandTypes.DOWNLOAD == CommandType)
+                {
+                    initFileMetadata(TransferMetadata.src_locations);
 
-            // Update the file metadata with GCS presigned URL
-            updatePresignedUrl();
+                    Directory.CreateDirectory(TransferMetadata.localLocation);
+                }
 
-            foreach (SFFileMetadata fileMetadata in FilesMetas)
-            {
-                // If the file is larger than the threshold, add it to the large files list
-                // Otherwise add it to the small files list
-                if (fileMetadata.srcFileSize > TransferMetadata.threshold)
+                // Update the file metadata with GCS presigned URL
+                updatePresignedUrl();
+
+                foreach (SFFileMetadata fileMetadata in FilesMetas)
                 {
-                    LargeFilesMetas.Add(fileMetadata);
+                    // If the file is larger than the threshold, add it to the large files list
+                    // Otherwise add it to the small files list
+                    if (fileMetadata.srcFileSize > TransferMetadata.threshold)
+                    {
+                        LargeFilesMetas.Add(fileMetadata);
+                    }
+                    else
+                    {
+                        SmallFilesMetas.Add(fileMetadata);
+                    }
                 }
-                else
+
+                // Check command type
+                if (CommandTypes.UPLOAD == CommandType)
                 {
-                    SmallFilesMetas.Add(fileMetadata);
+                    upload();
+                }
+                else if (CommandTypes.DOWNLOAD == CommandType)
+                {
+                    download();
                 }
             }
-
-            // Check command type
-            if (CommandTypes.UPLOAD == CommandType)
-            {
-                upload();
-            }
-            else if (CommandTypes.DOWNLOAD == CommandType)
+            catch (Exception e)
             {
-                download();
+                Logger.Error("Error while transferring file(s): " + e.Message);
+                if (e is SnowflakeDbException snowflakeException)
+                {
+                    if (snowflakeException.QueryId == null)
+                    {
+                        snowflakeException.QueryId = TransferMetadata.queryId;
+                    }
+                    throw snowflakeException;
+                }
+                throw new SnowflakeDbException(SFError.IO_ERROR_ON_GETPUT_COMMAND, TransferMetadata.queryId, e);
```

Comment:
Do we really need to override exceptions? The problem was with GetQueryId, not exception.
From user perspective isn't it better to see FileNotFoundException?

Reply:
Yes, but in JDBC @sfc-gh-dprzybysz implemented fixed the same problem in a way that both - exception and statement provide query id. I had change addressing exactly what you point out here (https://github.com/snowflakedb/snowflake-connector-net/pull/828/files) but I think drivers should be consistent.

Reply:
Actually we also need it to transport queryId to the SFStatement in case of a failure. Since this is transfer agent starting entire job with a query, to provide it on a statement layer we either have to return result.But in case of an exception there's the easiest way to achieve itwith  without wrapping the original exception and providing along with started and failed queryId. WDYT?